https://faang-school.atlassian.net/browse/BJS2-40664

ahim-miha opened the pull request:
Этапы проекта
1.
createStage(StageDto stageDto): Создает новый этап проекта.
 2.
getStageByFilter(StageFilterDto stageFilterDto): Получает список этапов, соответствующих заданным фильтрам. 
3.
deleteStage(Long stageId, ActionWithTask actionWithTask, Long transferStageId): Удаляет этап с возможностью каскадного удаления, закрытия или переноса задач. 
4.
updateStage(StageDto stageDto): Обновляет информацию об этапе, включая обеспечение необходимых ролей. 
5.
getAllProjectStages(Long projectId): Получает все этапы для указанного проекта. 
6.
getStageById(Long stageId): Получает информацию о конкретном этапе по его ID.
 7.
performCascadeDelete(Stage stage): Выполняет каскадное удаление этапа и связанных задач. 
8.
performCloseAction(Stage stage): Закрывает этап, отмечая все задачи как отмененные. 
9.
performTransferAction(Stage stage, Long transferStageId): Переносит задачи с одного этапа на другой. 
10.
ensureRequiredRoles(Stage stage, Map<TeamRole, Long> currentRoleCount, List<TeamMember> projectTeamMembers): Обеспечивает наличие необходимых ролей для этапа.

ahim-miha committed:
https://faang-school.atlassian.net/browse/BJS2-40664

Этапы проекта
1.
createStage(StageDto stageDto): Создает новый этап проекта.
2.
getStageByFilter(StageFilterDto stageFilterDto): Получает список этапов, соответствующих заданным фильтрам.
3.
deleteStage(Long stageId, ActionWithTask actionWithTask, Long transferStageId): Удаляет этап с возможностью каскадного удаления, закрытия или переноса задач.
4.
updateStage(StageDto stageDto): Обновляет информацию об этапе, включая обеспечение необходимых ролей.
5.
getAllProjectStages(Long projectId): Получает все этапы для указанного проекта.
6.
getStageById(Long stageId): Получает информацию о конкретном этапе по его ID.
7.
performCascadeDelete(Stage stage): Выполняет каскадное удаление этапа и связанных задач.
8.
performCloseAction(Stage stage): Закрывает этап, отмечая все задачи как отмененные.
9.
performTransferAction(Stage stage, Long transferStageId): Переносит задачи с одного этапа на другой.
10.
ensureRequiredRoles(Stage stage, Map<TeamRole, Long> currentRoleCount, List<TeamMember> projectTeamMembers): Обеспечивает наличие необходимых ролей для этапа.

Memoreks commented on src/main/java/faang/school/projectservice/controller/stage/StageController.java:
```diff
@@ -0,0 +1,49 @@
+@Slf4j
+@RequiredArgsConstructor
+@Controller
+public class StageController {
```
Я бы сразу добавил @RequestMapping

BobrAll replied:
на 7 спринте уже есть материалы по аннтациям RestController, RequestMapping, GetMapping, PostMapping, RequestBody и др., их бы сюда добавить сразу

Memoreks commented on src/main/java/faang/school/projectservice/controller/stage/StageController.java:
```diff
@@ -0,0 +1,49 @@
+
+    private final StageService stageService;
+
+    public StageDto createStage(StageDto stageDto){
```
И аннотации для эндпоинтов, в данном случае PostMapping

Memoreks commented on src/main/java/faang/school/projectservice/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,17 @@
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ControllerAdvice
+public class GlobalExceptionHandler {
```
Круто!

Memoreks commented on src/main/java/faang/school/projectservice/filter/stage/StageRolesFilter.java:
```diff
@@ -0,0 +1,25 @@
+
+import java.util.stream.Stream;
+
+@Controller
+public class StageRolesFilter implements StageFilter {
```
Почему аннотация контроллер?

Memoreks commented on src/main/java/faang/school/projectservice/service/stage/StageService.java:
```diff
@@ -0,0 +1,441 @@
+public class StageService {
+
+    private final StageRepository stageRepository;
+    private final ProjectRepository projectRepository;
+    private final StageInvitationRepository stageInvitationRepository;
```
Договорились, что напрямую в другие репозитории не обращаемся. Лучше переписать методы обращаться через сервисы, которые уже будут обращаться к репозиториям.

BobrAll replied:
Тут исправлений так и нет (по крайней мере к project-репозиторию обращение должно идти через сервис)

Memoreks commented on src/main/java/faang/school/projectservice/service/stage/StageService.java:
```diff
@@ -0,0 +1,441 @@
+        Stage stage = stageRepository.getById(stageId);
+        log.debug("Stage fetched successfully. Name: {}", stage.getStageName());
+
+        switch (actionWithTask) {
+            case CASCADE:
+                log.info("Performing CASCADE delete for stage: {}", stageId);
+                performCascadeDelete(stage);
+                break;
+            case CLOSE:
+                log.info("Performing CLOSE action for stage: {}", stageId);
+                performCloseAction(stage);
+                break;
+            case TRANSFER:
+                log.info("Performing TRANSFER action for stage: {} to stage: {}", stageId, transferStageId);
+                stageValidator.validationNumOnNullAndLessThanZero(transferStageId, "Transfer stage ID cannot be null or less than zero");
+                performTransferAction(stage, transferStageId);
+                break;
+            default:
+                log.error("Unknown task action: {}", actionWithTask);
+                throw new DataValidationException("Unknown task action: " + actionWithTask);
+        }
```
Для красоты можно попробовать переделать switch из java 12

Memoreks commented on src/test/java/faang/school/projectservice/service/stage/StageServiceTest.java:
```diff
@@ -0,0 +1,292 @@
+
+    @BeforeEach
+    public void setUp() {
+        stageRepository = mock(StageRepository.class);
+        projectRepository = mock(ProjectRepository.class);
+        stageInvitationRepository = mock(StageInvitationRepository.class);
+        stageValidator = mock(StageValidator.class);
+        stageFilter = mock(StageFilter.class);
+        stageMapper = spy(new StageMapperImpl());
+        filters = List.of(stageFilter);
+
+        stageService = new StageService(
+                stageRepository,
+                projectRepository,
+                stageInvitationRepository,
+                stageValidator,
+                stageMapper,
+                filters
+        );
```
Это все можно не писать если применить аннотации @Mock и @InjectMocks

Memoreks reviewed (commented):
Отличное решение. Особенно круто смотрится подробное логирование, лайк за это:)
Оставил небольшие комментарии

VovaBuzivskyi commented:
Код написан хорошо. Оставил несколько комментариев.

VovaBuzivskyi commented on src/main/java/faang/school/projectservice/dto/stage/StageDto.java:
```diff
@@ -0,0 +1,20 @@
+@Setter
+@ToString
+@Builder
+public class StageDto {
```
Здесь еще нужны конструкторы NoArgs AllArgs

VovaBuzivskyi commented on src/main/java/faang/school/projectservice/dto/stage/StageFilterDto.java:
```diff
@@ -0,0 +1,17 @@
+@Setter
+@ToString
+@Builder
+public class StageFilterDto {
+    private List<StageRoles> stageRoles;
```
Здесь тоже нужны конструкторы.  И я не думаю что это хорошая идея передавать внутри Дто фильтра ,сущности. Так как говорили передавать и возвращать с эндпоинтов только Дто а не сущности. 
Я думаю тебе стоит поменять фильтр , как я понял по заданию тебе нужно фильтровать по этим параметрам 
![Снимок экрана 2024-11-15 000435](https://github.com/user-attachments/assets/99f555a5-7a5f-47d9-8948-a0b442c0f739)
![Снимок экрана 2024-11-15 000311](https://github.com/user-attachments/assets/2e512b22-638d-4cb2-96eb-e2a60a6fe904)
 а не по списку сущностей StageRole.
T.e тебе нужно два фильтра типа String : teamRolePattern & taskStatusPattern

VovaBuzivskyi commented on src/main/java/faang/school/projectservice/filter/stage/StageRolesFilter.java:
```diff
@@ -0,0 +1,25 @@
+public class StageRolesFilter implements StageFilter {
+    @Override
+    public boolean isApplicable(StageFilterDto stageFilterDto) {
+        return stageFilterDto.getStageRoles() != null;
```
Если stageFilterDto будет null , получится NullPointerException . И не получится дойти до метода getStageRoles(). Нужна проверка самого фильтра на null

VovaBuzivskyi commented on src/main/java/faang/school/projectservice/filter/stage/StageRolesFilter.java:
```diff
@@ -0,0 +1,25 @@
+    }
+
+    @Override
+    public Stream<Stage> apply(Stream<Stage> stage, StageFilterDto stageFilterDto) {
```
В стриме по идее может быть много stages , а переменная называется stage (мелочь но лучше поменять)

VovaBuzivskyi commented on src/test/java/faang/school/projectservice/service/stage/StageServiceTest.java:
```diff
@@ -0,0 +1,292 @@
+    public void checkCreateStageSuccessTest() {
+        when(stageMapper.toEntity(stageDto))
+                .thenReturn(stage);
+        when(stageRepository.save(stage))
+                .thenReturn(stage);
```
Нету verify() вызова этого метода

VovaBuzivskyi commented on src/test/java/faang/school/projectservice/service/stage/StageServiceTest.java:
```diff
@@ -0,0 +1,292 @@
+                .thenReturn(stage);
+        when(stageRepository.save(stage))
+                .thenReturn(stage);
+        when(stageMapper.toDto(stage))
+                .thenReturn(stageDto);
```
Маппер Spy ,ему не нужно писать что возвращать, он сам вернет смапленный обьект.

VovaBuzivskyi commented on src/test/java/faang/school/projectservice/service/stage/StageServiceTest.java:
```diff
@@ -0,0 +1,292 @@
+
+        when(stageRepository.findAll())
+                .thenReturn(stageList);
+        when(filters.get(0).isApplicable(stageFilterDto))
+                .thenReturn(true);
+        when(filters.get(0).apply(any(), any()))
+                .thenReturn(stream);
```
Нету вызова verify() для методов: isApplicable и apply

VovaBuzivskyi commented on src/test/java/faang/school/projectservice/validtor/stage/StageValidatorTest.java:
```diff
@@ -0,0 +1,78 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+@ExtendWith(MockitoExtension.class)
```
Эта аннотация здесь не нужна , так как в этом классе нет Моков.

ahim-miha committed:
https://faang-school.atlassian.net/browse/BJS2-40664

Этапы проекта

Исправил замечания.
Добавил тесты к фильтрам.

BobrAll commented on src/main/java/faang/school/projectservice/filter/stage/StageTaskStatusFilter.java:
```diff
@@ -0,0 +1,23 @@
+import java.util.stream.Stream;
+
+public class StageTaskStatusFilter implements StageFilter {
+    @Override
```
У нас по код-стайлу принято вставлять пустые строки перед связкой ("аннотации + метод/поле"), можно глянуть в инфо дс

В остальных местах тоже поправить бы

BobrAll commented on src/main/java/faang/school/projectservice/filter/stage/StageTaskStatusFilter.java:
```diff
@@ -0,0 +1,23 @@
+    }
+
+    @Override
+    public Stream<Stage> apply(Stream<Stage> stages, StageFilterDto stageFilterDto) {
+        String taskStatusPattern = stageFilterDto.getTaskStatusPattern().toLowerCase();
+        return stages.filter(stage ->
+                stage.getTasks().stream()
+                        .anyMatch(task ->
+                                task.getStatus().name().toLowerCase().contains(taskStatusPattern)));
+    }
```
Круто

BobrAll commented on src/main/java/faang/school/projectservice/validator/stage/StageValidator.java:
```diff
@@ -0,0 +1,51 @@
+@Component
+public class StageValidator {
+
+    public <T> void validationOnNull(T value, String message) {
+        if (value == null) {
+            throw new DataValidationException(message);
+        }
+    }
+
+    public <T extends Number> void validationNumOnNullAndLessThanZero(T value, String message) {
+        if (value == null || value.doubleValue() <= 0) {
+            throw new DataValidationException(message);
+        }
+    }
```
Сильно!

BobrAll commented on src/test/java/faang/school/projectservice/service/stage/StageServiceTest.java:
```diff
@@ -0,0 +1,302 @@
+
+
+    @BeforeEach
+    public void setUp() {
```
Большеватый сетап, постарайся вынести из него по-максимуму все в методы (в таком случае контекст теста сразу будет виден)

BobrAll reviewed (commented):
Хороший код. Немного правок написал, потом самое главное - протестировать с postman все запросы, когда напишешь эндпоинты в контроллере

BobrAll commented:
@0Miha0 Скажи пожалуйста, почему ты не исправил (либо не прокомментировал, если мнения расходятся) замечания, которые тебе ребята на ревью написали?

BobrAll commented:
Также конфликт нужно поправить
<img width="959" alt="image" src="https://github.com/user-attachments/assets/68e0b671-f5b6-42a2-ac30-cbdbe7432824">

BobrAll commented:
> @0Miha0 Скажи пожалуйста, почему ты не исправил (либо не прокомментировал, если мнения расходятся) замечания, которые тебе ребята на ревью написали?

В дискорде видел обсуждение, если что - это круто, что ты там об этом спросил. Но изменений-то нет. По остальным комментариям нет ни обсуждений, ни изменений

BobrAll commented:
<img width="1128" alt="image" src="https://github.com/user-attachments/assets/e465d109-b98c-49eb-90fa-09c66601e70a">
Ага, я кажется понял. На ревью мне нужно отправлять уже со всеми исправлениями после код-ревью. Иначе получается, что мы делаем двойную работу - проверяем одно и то же

ahim-miha committed:
https://faang-school.atlassian.net/browse/BJS2-40664

Этапы проекта

Исправил замечания.

BobrAll commented on src/main/java/faang/school/projectservice/controller/stage/StageController.java:
```diff
@@ -0,0 +1,70 @@
+        return stageService.createStage(stageDto);
+    }
+
+    @GetMapping("/filter")
+    public List<StageDto> getStageWithFilter(@RequestBody StageFilterDto stageFilterDto) {
```
В спецификации не допускается тело в гет-запросе. Это единственное место, где для получения можно использовать метод post.
Все остальные эндпоинты - супер

BobrAll commented on src/main/java/faang/school/projectservice/controller/stage/StageController.java:
```diff
@@ -0,0 +1,70 @@
+@Slf4j
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/api/v1/stages")
```
В мастер-ветке уже есть приставка /api/v1, поэтому тут ее можно не делать (двойная тогда будет)
<img width="788" alt="image" src="https://github.com/user-attachments/assets/b19f77e0-9733-4bf3-9804-bd6e7e6158c6">

BobrAll commented on src/main/java/faang/school/projectservice/controller/stage/StageController.java:
```diff
@@ -0,0 +1,70 @@
+        return stageService.updateStage(stageDto);
+    }
+
+    @GetMapping("/project/{projectId}")
```
Тут бы во множественном числе сделать projects (так принято в рест для сущностей)

BobrAll reviewed (commented):
Объемная задача, но у тебя хорошо получилось с ней справиться. Еще пару комментариев оставил, потом можно вливать в мастер.

ahim-miha committed:
https://faang-school.atlassian.net/browse/BJS2-40664

Этапы проекта

Исправил замечания.
И если ты не против, можно я тесты трогать не буду?

ahim-miha committed:
https://faang-school.atlassian.net/browse/BJS2-40664

Этапы проекта

Исправил замечания.
И если ты не против, можно я тесты трогать не буду??

BobrAll commented on src/main/java/faang/school/projectservice/controller/stage/StageController.java:
```diff
@@ -0,0 +1,70 @@
+@Slf4j
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("stages")
```
stages в RequestMapping должно начинаться с / 
Тестировал запросы?

ahim-miha replied:
Да, все работает, теперь точно работает

BobrAll reviewed (changes requested):
Тесты все еще падают 
<img width="948" alt="image" src="https://github.com/user-attachments/assets/86538ff9-10c2-490a-8b83-142c4c5d5a2f">

ahim-miha committed:
https://faang-school.atlassian.net/browse/BJS2-41118

Поправил все, немного переделпл метод удаления, снес логи, ибо мешают очень, протестил через постман, переделал тесты, ибо перестали работать, все работает(наверно)

ahim-miha committed:
Merge branch 'refs/heads/Manticore-master-stream7' into feature-BJS2-40664

# Conflicts:
#	src/main/java/faang/school/projectservice/exception/DataValidationException.java
#	src/main/java/faang/school/projectservice/exception/ErrorResponse.java
#	src/main/java/faang/school/projectservice/exception/GlobalExceptionHandler.java
#	src/main/java/faang/school/projectservice/exception/MessageError.java
#	src/main/java/faang/school/projectservice/service/project/ProjectService.java
#	src/main/java/faang/school/projectservice/service/stage/StageService.java
#	src/test/java/faang/school/projectservice/service/stage/StageServiceTest.java

ahim-miha committed:
https://faang-school.atlassian.net/browse/BJS2-41118

Поправил тесты

ahim-miha committed:
https://faang-school.atlassian.net/browse/BJS2-41118

Поправил тесты

BobrAll reviewed (approved):
Отлично, вливаю
